### ISSUE TITLE
ID3 Tags not displayed /ignored

### ISSUE THREAD

**Issue by yafp** [clementine-player/Clementine#5946]
### System information

- Operating System: Arch Linux
- Clementine version:  1.3.1

### Expected behaviour / actual behaviour
- I am editing ID3/meta-informations via clementine starting from a playlist
- I expect to see the inserted information after SAVEing in the playlist

but what it does
- still shows the old meta information in the playlist
- and in addition: the library view is not updated according to the entered informations

If i reopen editing of meta-informations of that file the updated strings are displayed inside the editing window

### Steps to reproduce the problem (only for bugs)
- can't tell if that is reproducible for others

**Comment by aelzenaar** [comment]
I'm running a git build that I compiled manually a few weeks ago on Arch, I do not have this problem. Does this problem still exist for you if you install the clementine-git AUR package?

**Comment by yafp** [comment]
I did a test run using the clementine-git from AUR with the same result. So yes - the problem still exist for me with the version 1.3.1 from AUR.

**Comment by Moonbase59** [comment]
I can confirm this for the following v1.3.1 builds:

- Windows 7
- Ubuntu 14.04.5
- Linux Mint 17.3

**Comment by yafp** [comment]
@Moonbase59 
thanks for confirming this bug. until now i was unsure if its a specific problem with my setup.

**Comment by nevvious** [comment]
Also confirm (v.1.3.1, Arch Linux).

**Comment by kunver400** [comment]
confirmed for Ubuntu 17.01 too.

**Comment by Bodale75** [comment]
Also confirm in Debian 8.10, Mate 1.8.1, Clementine Vers. 1.2.3 (installed from repo).

**Comment by logological** [comment]
Possibly a duplicate of #5144?

**Comment by thenewnano** [comment]
Confirmed 
Changes are made to file but not to database, full library scan is extremely slow in Mac and absolutely not an option, creating new playlists doesn't help, tags are not changed in the UI or library View regardless. Same version of Clementine in Ubuntu works just fine!

clementine-1.3.1-537-g3c8a558
osx 10.14.3 Beta (18D32a)
Music Library with over 40000 files

**Comment by thenewnano** [comment]
Did some digging, my theory is that when the Edit Tag Menu action "save" is invoked a function loops trhough all the data comparing to the orginal values and if changed then  
 SaveFileBlocking(const QString& filename, const Song& metadata); is called, this will update the file. 

what happens next that has made this bug a bit of a ghost is the updated file might in some systems/installation/library size invoke a rescan of changed files which in turn updates the database and runs trough all functions triggered back to libraryview UI and updates the lib list.
But in some other cases the underlaying filesystem watcher never becomes aware of the change or fails to handle that and no chain of updating the database is started. 

I guess when the SaveFileBlocking is called and returns true we should call update database / tags etc for that sepcific file and go back to the previous UI. Obviously this might in some situation invoke two times of reread of the file, but it shouldn't be an issue as the second read will not do much as the data is already there. if the force read also is blocking then we should be safe. 

I guess I need to digg a bit deeper to before I can put together a patch and then make sure I can compile and all . would be really great if one of the contributors could just try to call update database on changed file and confirm it doesn't cause any issues.

**Comment by thenewnano** [comment]
So I did compile on my Ubuntu machine and went trough the logical path for file changes, finding that every time I save changes in the tag-editor a library scan on the changed file is triggered, as it should. 
I admit it seems that change reloads are invoked more than once, which could be improved but I don't want to go after that :) 
`16:48:17.076 DEBUG TagReader:679                    Saving tags to "/home/....mp3"
16:48:20.948 DEBUG LibraryWatcher:356               "/home/....mp3" changed 
16:48:20.949 DEBUG _MessageReplyBase:24             Waiting on ID 3 
16:48:20.949 DEBUG TagReader:132                    Reading tags from "/home/...mp3" 
16:48:20.949 DEBUG TagReader:488                    Parsing FMPSFrame "FMPS_PlayCount" ,  "0" 
16:48:20.950 DEBUG TagReader:488                    Parsing FMPSFrame "FMPS_Rating_Amarok_Score" ,  "0" 
16:48:20.950 DEBUG MessageReply<MessageType>:90     Releasing ID 3 (finished) 
16:48:20.950 DEBUG _MessageReplyBase:26             Acquired ID 3 
16:48:20.950 DEBUG LibraryWatcher:553               "/home/....mp3" metadata changed 
16:48:21.938 DEBUG LibraryWatcher:620               Subdir "/home/..." changed under directory "/home/..." id 1 
16:48:21.938 DEBUG LibraryWatcher:620               Subdir "/home/..." changed under directory "/home/..." id 1 
16:48:21.965 DEBUG LibraryWatcher:620               Subdir "/home/..." changed under directory "/home/..." id 1 
16:48:21.965 DEBUG LibraryWatcher:620               Subdir "/home/..." changed under directory "/home/..." id 1 
16:48:26.156 DEBUG LibraryWatcher:620               Subdir "/home/..." changed under directory "/home/..." id 1 
16:48:26.156 DEBUG LibraryWatcher:620               Subdir "/home/..." changed under directory "/home/... id 1 
16:48:26.177 DEBUG LibraryWatcher:620               Subdir "/home/..." changed under directory "/home/..." id 1 
16:48:26.177 DEBUG LibraryWatcher:620               Subdir "/home/..." changed under directory "/home/..." id 1 
16:49:10.758 DEBUG Database:600                     Starting database integrity check 
16:49:11.286 DEBUG Database:661                     Starting database backup 
16:49:13.468 DEBUG LibraryWatcher:620               Subdir "/home/..." changed under directory "/home/..." id 1 
16:49:13.468 DEBUG LibraryWatcher:620               Subdir "/home/..." changed under directory "/home/..." id 1 
16:49:13.485 DEBUG LibraryWatcher:620               Subdir "/home/..." changed under directory "/home/..." id 1 
`

 well well this seems to have been a known issue at least to @hatstand  on Linux  for like five years ago as found in tagreader.cpp
`#ifdef Q_OS_LINUX
  if (ret) {
    // Linux: inotify doesn't seem to notice the change to the file unless we
    // change the timestamps as well. (this is what touch does)
    utimensat(0, QFile::encodeName(filename).constData(), nullptr, 0);
  }
#endif  // Q_OS_LINUX

  return ret;
}`

The work around was to force change time stamps when a file was changed, I guess as we do find some similar issue on Windows as well as OSX , in the latest OSX installations it seems to always be the case, I have tried case sensitive and case insensitive file systems , so I suggest we remove the compiler flag here and do this step for all OSs . I'm going to try to cross compile now

**Comment by thenewnano** [comment]
So I had to change to utime() to make it work in OSX without much fiddle and code change. Now we are basically touching the file, that is all good, maybe it was all good before my magical changed and compile pain for OSX. the finding now is that the folder including the files has to change for the watcher to trigger. so what I did doing a simple utime() on the edited filed did nothing, Clementine is watching folders and if a file is changed it goes trough checking all files in that folder.  I'll see if I can patch to do a utime on the folder instead, getting closer .

**Comment by thenewnano** [comment]
So I did some improvements and sorted my changes a bit better, now I force change date on the folder as well, and it sometimes results in updated database but mostly not.  tried to automatically kick in a incremental scan on save .. but didn't manage.
[fix-dir-scan.txt](https://github.com/clementine-player/Clementine/files/2719069/fix-dir-scan.txt)

This patch solves big part of the issue as when file changes are saved:
- Update changed library Folders  finds the new tags 
- Restarting Clementine results the same as above and new tags
- The time based changed folder scan will catch the changeI think it was 1000 seconds intervals

**Comment by yafp** [comment]
any chance this patch gets included?
